tools/extract_baserom.c

```diff
@@ -0,0 +1,1604 @@
+    "mahouya_scene",
+    "mahouya_room_0",
+    "takaraya_scene",
+    "takaraya_room_0",
+    "takaraya_room_1",
+    "takaraya_room_2",
+    "takaraya_room_3",
+    "takaraya_room_4",
+    "takaraya_room_5",
+    "takaraya_room_6",
+    "ganon_sonogo_scene",
+    "ganon_sonogo_room_0",
+    "ganon_sonogo_room_1",
+    "ganon_sonogo_room_2",
+    "ganon_sonogo_room_3",
+    "ganon_sonogo_room_4",
+    "ganon_demo_scene",
+    "ganon_demo_room_0",
+    "besitu_scene",
+    "besitu_room_0",
+    "face_shop_scene",
+    "face_shop_room_0",
+    "kinsuta_scene",
+    "kinsuta_room_0",
+    "ganontikasonogo_scene",
+    "ganontikasonogo_room_0",
+    "ganontikasonogo_room_1",
+    "test01_scene",
+    "test01_room_0",
+    "bump_texture_static",
+    "anime_model_1_static",
+    "anime_model_2_static",
+    "anime_model_3_static",
+    "anime_model_4_static",
+    "anime_model_5_static",
+    "anime_model_6_static",
+    "anime_texture_1_static",
+    "anime_texture_2_static",
+    "anime_texture_3_static",
+    "anime_texture_4_static",
+    "anime_texture_5_static",
+    "anime_texture_6_static",
+    "softsprite_matrix_static",
+};
+
+int main(void) {
+    FILE* rom = fopen("baserom.z64", "rb");
+    FILE* outFile;
+    unsigned int virtualStart;
+    unsigned int virtualEnd;
+
+    if (rom == NULL) {
+        fprintf(stderr, "Could not open baserom.z64\n");
+        exit(1);
+    }
+
+    for (unsigned int i = 0; i < DMA_TABLE_ENTRIES; i++) {
+        char outFileName[45] = "baserom/";
+        fseek(rom, DMA_TABLE_OFFSET + (DMA_ENTRY_SIZE * i), SEEK_SET);
+
+        if (fread(&virtualStart, sizeof(unsigned int), 1, rom) != 1) {
+            fprintf(stderr, "Could not read start address for file %s\n", sDmaMgrFileNames[i]);
+            exit(1);
+        }
+
+        virtualStart = htobe32(virtualStart);
+
+        if (fread(&virtualEnd, sizeof(unsigned int), 1, rom) != 1) {
+            fprintf(stderr, "Could not read end address for file %s\n", sDmaMgrFileNames[i]);
+            exit(1);
+        }
+
+        virtualEnd = htobe32(virtualEnd);
+
+        void* data = malloc(virtualEnd - virtualStart);
+
+        outFile = fopen(strcat(outFileName, sDmaMgrFileNames[i]), "wb");
+
+        if (data == NULL) {
+            fprintf(stderr, "Could not allocate memory for %s\n", outFileName);
+            exit(1);
+        }
+
+        if (outFile == NULL) {
+            fprintf(stderr, "Could not open output file for %s\n", outFileName);
+            exit(1);
+        }
+
+        fseek(rom, virtualStart, SEEK_SET);
+
+        if (fread(data, virtualEnd - virtualStart, 1, rom) != 1) {
+            fprintf(stderr, "Could not read rom file %s\n", sDmaMgrFileNames[i]);
+        }
+
+        printf("Extracting file %s	0x%x - 0x%x\n", sDmaMgrFileNames[i], virtualEnd, virtualStart);
+
+        if (fwrite(data, virtualEnd - virtualStart, 1, outFile) != 1) {
+            fprintf(stderr, "Could not write rom file %s\n", sDmaMgrFileNames[i]);
+        }
```

Comment:
I was thinking, do we really need to read the file and write it again?
Maybe there's a way to write a portion of the rom to a new file without actually moving it to RAM. I remember reading something about `mmap` was used for stuff like this, but I don't really remember.

Reply:
From what I just looked up `mmap` is just wrapped by `malloc` and on top of that ts a system function which could make more of a mess.